File: service/src/whf/ai/usage.py
"""What one Copilot narration cost, in one shape whatever the source.

Two sources answer the question, and they answer it differently. The session's own metrics
(`session.rpc.usage.get_metrics()`) are what the account is billed: AI credits, the legacy
premium-request cost, the time spent in the model API. The `assistant.usage` events the session
streams carry tokens per call and nothing about money. Both are folded into the same dict here, with
`source` saying which one it came from, so the CLI and the app can show what is known and stay quiet
about what is not: every value is `None` when it is unknown, never a plausible-looking zero.

Since 1 June 2026 Copilot bills AI credits; one credit is one US cent, and the SDK reports credits
as nano-AI units (credits × 1e9).
"""

from __future__ import annotations

from typing import Any

TOKEN_FIELDS = ("input_tokens", "output_tokens", "cache_read_tokens", "reasoning_tokens")
NANO_PER_CREDIT = 1e9
CREDITS_PER_USD = 100.0


def empty_usage() -> dict:
    """Nothing is known: no session was ever created, so no number can be honest.

    `models` is the one exception to "unknown is None": it is a mapping the app iterates over, so
    the empty mapping is both true and easier to consume than a null.
    """
    return {
        "input_tokens": None,
        "output_tokens": None,
        "cache_read_tokens": None,
        "reasoning_tokens": None,
        "requests": None,
        "premium_requests": None,
        "ai_credits": None,
        "usd": None,
        "api_seconds": None,
        "models": {},
        "source": "none",
    }


def usage_from_metrics(metrics: Any) -> dict:
    """The session's billed usage, from an SDK `UsageGetMetricsResult`."""
    model_metrics: dict[str, Any] = dict(getattr(metrics, "model_metrics", None) or {})
    nano_aiu = getattr(metrics, "total_nano_aiu", None)
    credits = None if nano_aiu is None else float(nano_aiu) / NANO_PER_CREDIT
    return {
        "input_tokens": sum(int(m.usage.input_tokens) for m in model_metrics.values()),
        "output_tokens": sum(int(m.usage.output_tokens) for m in model_metrics.values()),
        "cache_read_tokens": sum(int(m.usage.cache_read_tokens) for m in model_metrics.values()),
        # A model that does not reason reports no reasoning tokens; it did use zero of them.
        "reasoning_tokens": sum(int(m.usage.reasoning_tokens or 0) for m in model_metrics.values()),
        "requests": int(metrics.total_user_requests),
        "premium_requests": float(metrics.total_premium_request_cost),
        "ai_credits": credits,
        "usd": None if credits is None else credits / CREDITS_PER_USD,
        "api_seconds": float(metrics.total_api_duration_ms) / 1000.0,
        "models": {
            name: {
                "requests": int(m.requests.count),
                "input_tokens": int(m.usage.input_tokens),
                "output_tokens": int(m.usage.output_tokens),
            }
            for name, m in model_metrics.items()
        },
        "source": "metrics",
    }


def usage_from_events(events: list[dict]) -> dict:
    """The tokens the streamed `assistant.usage` events reported, one event per model call.

    A field no event carried stays `None`: the events simply did not say. Money is never known
    here, so it stays `None` too, which is what tells the app to show tokens only.
    """
    usage = empty_usage()
    for field in TOKEN_FIELDS:
        reported = [event[field] for event in events if event.get(field) is not None]
        usage[field] = sum(int(value) for value in reported) if reported else None
    usage["requests"] = len(events)
    models: dict[str, dict] = {}
    for event in events:
        name = event.get("model")
        if name is None:  # an event that does not name its model cannot be attributed to one
            continue
        entry = models.setdefault(name, {"requests": 0, "input_tokens": 0, "output_tokens": 0})
        entry["requests"] += 1
        entry["input_tokens"] += int(event.get("input_tokens") or 0)
        entry["output_tokens"] += int(event.get("output_tokens") or 0)
    usage["models"] = models
    usage["source"] = "events"
    return usage

File: server/forecast-core/src/main/java/com/workloadhub/forecast/ai/Usage.java
package com.workloadhub.forecast.ai;

import com.workloadhub.forecast.data.ExportFiles;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/**
 * What one narration cost, in one shape whatever the source. The session metrics are the billed truth (credits,
 * premium requests, API time); the streamed events carry tokens only. Unknown is null, never a plausible zero;
 * {@code models} is the one exception (an empty map). One AI credit is one US cent; the SDK reports nano-AI units.
 */
final class Usage {

    static final double NANO_PER_CREDIT = 1e9;
    static final double CREDITS_PER_USD = 100.0;

    private Usage() {
    }

    static Map<String, Object> empty() {
        Map<String, Object> u = new LinkedHashMap<>();
        u.put("input_tokens", null);
        u.put("output_tokens", null);
        u.put("cache_read_tokens", null);
        u.put("reasoning_tokens", null);
        u.put("requests", null);
        u.put("premium_requests", null);
        u.put("ai_credits", null);
        u.put("usd", null);
        u.put("api_seconds", null);
        u.put("models", new LinkedHashMap<String, Object>());
        u.put("source", "none");
        return u;
    }

    static Map<String, Object> fromMetrics(UsageMetrics m) {
        Map<String, Object> u = empty();
        long input = 0;
        long output = 0;
        long cache = 0;
        long reasoning = 0;
        Map<String, Object> models = new LinkedHashMap<>();
        for (Map.Entry<String, UsageMetrics.ModelMetric> e : m.models().entrySet()) {
            UsageMetrics.ModelMetric mm = e.getValue();
            input += mm.inputTokens();
            output += mm.outputTokens();
            cache += mm.cacheReadTokens();
            reasoning += mm.reasoningTokens() == null ? 0 : mm.reasoningTokens();
            Map<String, Object> entry = new LinkedHashMap<>();
            entry.put("requests", mm.requests());
            entry.put("input_tokens", mm.inputTokens());
            entry.put("output_tokens", mm.outputTokens());
            models.put(e.getKey(), entry);
        }
        Double credits = m.totalNanoAiu() == null ? null : m.totalNanoAiu() / NANO_PER_CREDIT;
        u.put("input_tokens", input);
        u.put("output_tokens", output);
        u.put("cache_read_tokens", cache);
        u.put("reasoning_tokens", reasoning);
        u.put("requests", m.totalUserRequests());
        u.put("premium_requests", m.totalPremiumRequestCost());
        u.put("ai_credits", credits);
        u.put("usd", credits == null ? null : credits / CREDITS_PER_USD);
        u.put("api_seconds", m.totalApiDurationMs() == null ? null : m.totalApiDurationMs() / 1000.0);
        u.put("models", models);
        u.put("source", "metrics");
        return u;
    }

    static Map<String, Object> fromEvents(List<UsageEvent> events) {
        Map<String, Object> u = empty();
        u.put("input_tokens", sum(events, UsageEvent::inputTokens));
        u.put("output_tokens", sum(events, UsageEvent::outputTokens));
        u.put("cache_read_tokens", sum(events, UsageEvent::cacheReadTokens));
        u.put("reasoning_tokens", sum(events, UsageEvent::reasoningTokens));
        u.put("requests", (long) events.size());
        Map<String, Object> models = new LinkedHashMap<>();
        for (UsageEvent e : events) {
            if (e.model() == null) {
                continue;
            }
            @SuppressWarnings("unchecked")
            Map<String, Object> entry = (Map<String, Object>) models.computeIfAbsent(e.model(), k -> {
                Map<String, Object> fresh = new LinkedHashMap<>();
                fresh.put("requests", 0L);
                fresh.put("input_tokens", 0L);
                fresh.put("output_tokens", 0L);
                return fresh;
            });
            entry.put("requests", (Long) entry.get("requests") + 1);
            entry.put("input_tokens", (Long) entry.get("input_tokens") + (e.inputTokens() == null ? 0 : e.inputTokens()));
            entry.put("output_tokens", (Long) entry.get("output_tokens") + (e.outputTokens() == null ? 0 : e.outputTokens()));
        }
        u.put("models", models);
        u.put("source", "events");
        return u;
    }

    /** The sum of the events that carried the field, or null when none did. */
    private static Long sum(List<UsageEvent> events, Function<UsageEvent, Long> field) {
        Long total = null;
        for (UsageEvent e : events) {
            Long v = field.apply(e);
            if (v != null) {
                total = (total == null ? 0 : total) + v;
            }
        }
        return total;
    }

    static String toJson(Map<String, Object> usage) {
        return ExportFiles.mapper().writeValueAsString(usage);
    }
}
